patches/server/1036-Add-predicate-for-blocks-when-raytracing.patch

```diff
@@ -0,0 +1,120 @@
+index 0e670de77a7f9926e295e1dd63d909bed1a959ca..4e1822a94c0e52654d6afcb3bb6236b3098b6459 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -1117,9 +1117,17 @@ public class CraftWorld extends CraftRegionAccessor implements World {
+ 
+     @Override
+     public RayTraceResult rayTraceEntities(Location start, Vector direction, double maxDistance, double raySize, Predicate<Entity> filter) {
++        // Paper Start
++        return rayTraceEntities((io.papermc.paper.math.Position) start, direction, maxDistance, raySize, filter);
++    }
++
++    public RayTraceResult rayTraceEntities(io.papermc.paper.math.Position start, Vector direction, double maxDistance, double raySize, Predicate<Entity> filter) {
+         Preconditions.checkArgument(start != null, "Location start cannot be null");
+-        Preconditions.checkArgument(this.equals(start.getWorld()), "Location start cannot be in a different world");
+-        start.checkFinite();
++        Preconditions.checkArgument(!(start instanceof Location location) || this.equals(location.getWorld()), "Location start cannot be in a different world");
++        org.bukkit.util.NumberConversions.checkFinite(start.x(), "x not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.y(), "y not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.z(), "z not finite");
++        // Paper End
+ 
+         Preconditions.checkArgument(direction != null, "Vector direction cannot be null");
+         direction.checkFinite();
+@@ -1169,9 +1177,18 @@ public class CraftWorld extends CraftRegionAccessor implements World {
+ 
+     @Override
+     public RayTraceResult rayTraceBlocks(Location start, Vector direction, double maxDistance, FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks) {
++        // Paper start
++        return this.rayTraceBlocks(start, direction, maxDistance, fluidCollisionMode, ignorePassableBlocks, null);
++    }
++
++    @Override
++    public RayTraceResult rayTraceBlocks(io.papermc.paper.math.Position start, Vector direction, double maxDistance, FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks, Predicate<Block> passThrough) {
+         Preconditions.checkArgument(start != null, "Location start cannot be null");
+-        Preconditions.checkArgument(this.equals(start.getWorld()), "Location start cannot be in a different world");
+-        start.checkFinite();
++        Preconditions.checkArgument(!(start instanceof Location location) || this.equals(location.getWorld()), "Location start cannot be in a different world");
++        org.bukkit.util.NumberConversions.checkFinite(start.x(), "x not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.y(), "y not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.z(), "z not finite");
++        // Paper end
```

Comment:
Go ahead and at a `boolean isFinite` method to `Position` which the default implementation just checks if each component x, y, z are finite and returns true if it is. Then override that method in Location to also check the pitch and yaw. Then if its not, you can throw illegal argument exception.

Reply:
Added.